# Contents
# 1. Import libraries
# 2. Importing Data
# 3. Mixed-type Data
# 4. Missing Values
# 5. Addressing missing values
## 5.1 Exkurs: when using the median, df['column with missings'].fillna(median value, inplace=True)
# 6. Duplicates
# 7. Export changes
# 8. Task 4.5
## 8.1 Task 2. Run df.describe() for df_ords, interpret the output..
## 8.2 Step 3. Check for mixed-type data in df_ords dataframe.
## 8.3 Step 4. If mixed-type data finded, fix it. - No mixed-type data found.
## 8.4 Step 5. Run a check for missing values in df_ords dataframe. Report findings and propose an explanation for any missing values finded.
## 8.5 Step 6. Address the missing values using an appropriate method. Explain why you used your method of choice.
### 8.5.1 Imputing made correctly
## 8.6 Step 7. Run a check for duplicate values in your df_ords data. Report findings and propose an explanation for any duplicate values find.
## 8.7 Step 8. Address the duplicates using an appropriate method, explain why you used your method of choice.
## 8.8 Step 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

# 1. Import libraries

In [4]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [6]:
# Turn project folder path into a string
path=r'/Users/yevgeniyaem/Documents/Weiterbildung Data Analytics/11-2024 Instacart Basket Analysis'

In [8]:
path

'/Users/yevgeniyaem/Documents/Weiterbildung Data Analytics/11-2024 Instacart Basket Analysis'

In [10]:
# Import "orders.csv"
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [12]:
# Import "products.csv"
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 3. Mixed-type Data

### Instacart data is already preped so we will practice fixing mixed-type data by creating a small test dataframe

In [14]:
# Create a test df
df_test = pd.DataFrame()

In [18]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [20]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [22]:
# Check for any mixed type columns 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [24]:
# Convert column's data from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

In [26]:
# Checking results
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


# 4. Missing Values

In [28]:
# Find missing values (takes sum of Trues in a column)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [30]:
# Create a subset of the dataframe containing missing values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [32]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 5. Addressing missing values

### There a few ways to deal with missing data:
1. Create a new variable that acts like a flag based on the missing value.
2. Impute the value with the mean or median of the column (if the variable is numeric).
3. Remove or filter out the missing data.

In [34]:
# Print statistics
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

### Ex: When using the mean, df['column with missings'].fillna(mean value, inplace=True)

In [36]:
df_nan.median()

product_id       2413.0
product_name        NaN
aisle_id           98.5
department_id      11.5
prices            13.65
dtype: object

## 5.1 Exkurs: when using the median, df['column with missings'].fillna(median value, inplace=True)

### Because the missing values are strings there's not much you can do other than remove/filter the data

In [38]:
# Find shape of df
df_prods.shape

(49693, 5)

In [40]:
# Create new df without the nulls
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [42]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [44]:
# Checking: should have exactly 16 less rows
df_prods_clean.shape

(49677, 5)

### To drop all missing values --> df_prods.dropna(inplace = True)

### To drop only the NaNs from a particular column --> df_prods.dropna(subset = [‘product_name’], inplace = True)

# 6. Duplicates

In [46]:
# Create subset of df_prods_clean that contains only rows of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [48]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [50]:
# Check current number of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [52]:
# Creating new df without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [54]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [56]:
# Check current number of rows in df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

# 7. Export changes

In [58]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [60]:
df_prods_clean_no_dups.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 8. Task 4.5

## 8.1 Step 2. Run df.describe() for df_ords, interpret the output, share whether anything about the data looks off or should be investigated further.

In [62]:
df_ords.describe()

Unnamed: 0      order_id       user_id  customers_order_count  \
count  3.421083e+06  3.421083e+06  3.421083e+06           3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05           1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04           1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00           1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04           5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05           1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05           2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05           1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### "order_id", "user_id" - should be considered string values instead of numerical, bc there are all unique numbers.
### "customers_order_count" - has a max value of 100, which may mean orders are capped at 100 orders per customer.
### "orders_day_of_week" - has min value of 0 and max value of 6 meaning the seven numbers (0-6) represent days of the week.¶
### "order_hour_of_day" - has a min value of 0 and max value of 23 meaning the twenty-four numbers (0-23) represent 24 hours of the day.
### "days_since_prior_order" - has a count less than all other rows, what indicates missing values. It also indicates that some orders could be places within the same day. The max value is 30, which may mean that orders are capped at 30 days or 1 month.

In [64]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   customers_order_count   int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [66]:
# Change data type in "order_id"to string values instead of numerical.
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [68]:
# Change data type in "user_id"to string values instead of numerical.
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [70]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                object 
 2   user_id                 object 
 3   customers_order_count   int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(2)
memory usage: 182.7+ MB


## 8.2 Step 3. Check for mixed-type data in df_ords dataframe.

In [72]:
# Check for any mixed type columns 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### No columns with mixed type data identified

## 8.3 Step 4. If mixed-type data finded, fix it. - No mixed-type data found.

In [74]:
# Check the columns
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                object 
 2   user_id                 object 
 3   customers_order_count   int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(2)
memory usage: 182.7+ MB


## 8.4 Step 5. Run a check for missing values in df_ords dataframe. Report findings and propose an explanation for any missing values finded.

In [76]:
# Find missing values (takes sum of Trues in a column)
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
customers_order_count          0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### The "days_since_prior_order" column are missing 206209 values. These are probably first orders placed by customers. An explanation for this quanitty of missing values could be customers' first orders don't have a value. (acc.data dictionary Instacart has 206k users, what mathes with missing value number) This is relevant information that can't be removed, it should be preserved.

## 8.5 Step 6. Address the missing values using an appropriate method. Explain why you used your method of choice.

### First i'll verify if the missing values in the days_since_prior_order column correspond to rows where customers_order_count = 1, we can use a simple filtering approach. 
### I will use Imputation method with a Placeholder Value. I'll fill the missing values in the days_since_prior_order column with a placeholder value of 0. This approach is appropriate because:
### Missing values in this column are associated with the first orders placed by customers, where there is no prior order to calculate a difference.
### Using 0 preserves the information that these are first-time orders while avoiding confusion with subsequent orders. This is simple, logical, and consistent with the interpretation of the column.

In [78]:
# Check rows with missing values in 'days_since_prior_order'
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()]

In [80]:
df_ords_nan

Unnamed: 0 order_id user_id  customers_order_count  \
0                 0  2539329       1                      1   
11               11  2168274       2                      1   
26               26  1374495       3                      1   
39               39  3343014       4                      1   
45               45  2717275       5                      1   
...             ...      ...     ...                    ...   
3420930     3420930   969311  206205                      1   
3420934     3420934  3189322  206206                      1   
3421002     3421002  2166133  206207                      1   
3421019     3421019  2227043  206208                      1   
3421069     3421069  3154581  206209                      1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 7 columns]

In [84]:
# Verify if all these rows have 'order_number = 1'
all_first_orders = df_ords_nan['customers_order_count'].eq(1).all()

In [86]:
# Print result
if all_first_orders:
    print("All missing values in 'days_since_prior_order' belong to first orders (customers_order_count = 1).")
else:
    print("Not all missing values in 'days_since_prior_order' belong to first orders.")

All missing values in 'days_since_prior_order' belong to first orders (customers_order_count = 1).


In [88]:
# Impute missing values only for rows where 'customers_order_count' is 1
df_ords.loc[
    (df_ords['days_since_prior_order'].isnull()) & (df_ords['customers_order_count'] == 1), 
    'days_since_prior_order'
] = 0

In [90]:
# Verify if the missing values have been imputed correctly
print(df_ords['days_since_prior_order'].isnull().sum())  # Should now exclude rows where 'customers_order_count' = 1

0


In [92]:
# Check rows with missing values in 'days_since_prior_order'
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()]

df_ords_nan

### 8.5.1 Imputing made correctly

In [94]:
# Check the df
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                object 
 2   user_id                 object 
 3   customers_order_count   int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(2)
memory usage: 182.7+ MB


## 8.6 Step 7. Run a check for duplicate values in your df_ords data. Report findings and propose an explanation for any duplicate values find.

In [96]:
# Create subset of df_ords_clean that contains only rows of duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [97]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, customers_order_count, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### The dataframe created to check duplicates has returned empty, meaning there are no duplicates in the data set

## 8.7 Step 8. Address the duplicates using an appropriate method, explain why you used your method of choice.

### No duplicates foud in df_ords

## 8.8 Step 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [100]:
# Cleaned df_prods already exported as products_checked.csv
# Cleaned df_ords as orders_checked.csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [101]:
df_prods.shape

(49693, 5)

In [104]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3